### ISSUE TITLE
lxc-execute doesn't work with lxc.rootfs.options=ro

### ISSUE THREAD

**Issue by flx42** [lxc/lxc#2339]
On [current master](f49098e0d3b2dc7162ee0939b8282e255405156c), I get the following behavior:
```
$ lxc-create -t download -n xenial -- --dist ubuntu --release xenial --arch amd64

$ lxc-execute --define lxc.rootfs.options=ro -n xenial -- ls /
lxc-execute: xenial: conf.c: lxc_execute_bind_init: 3245 Read-only file system - Failed to create dummy "/usr/lib/lxc/rootfs/tmp/.lxc-init" file as bind mount target
lxc-execute: xenial: conf.c: lxc_setup: 3398 Failed to bind-mount the lxc init system
```
Looks like the rootfs was already remounted read-only at this point:
https://github.com/lxc/lxc/blob/f49098e0d3b2dc7162ee0939b8282e255405156c/src/lxc/conf.c#L3240-L3246

I see a few options:
1. Move call to `lxc_execute_bind_init` to `do_rootfs_setup`, after the `pre-mount` hooks.
It will require modifying the interface of `do_rootfs_setup` which is an exported function (internally).
This is probably the simplest.

2. Mount a new `tmpfs` directory, _e.g._ `/tmp/.lxc`, bind-mount the init inside it.

3. Use the only tmpfs at our disposal: `/dev`, seems like it's what docker does, they bind-mount to `/dev/init`. However, given the commit message in 794248d09d3a60633c901becc9d985db697f2feb, it's probably not possible for the same reason.

4. Use fexecve for the init (if that's possible).

Thoughts @brauner?

**Comment by brauner** [comment]
@tych0 had a patch to use `fexecve()` and we discussed it. The problem is that it's not supported on kernels we still support. Ideally I'd like a solution to use `fexecve()` if available otherwise fallback to using a bind-mount but we had no clear idea how to do it. Might be simplest to just do 1.

**Comment by tych0** [comment]
It's not that it's not clear how to do it, testing for fexecve is pretty easy. I just gave up because I already had workaround code for the problem I was having with bind-mounting init in. That said, it would let this case work on newer kernels, so perhaps it's worth resurrecting the patch? We could just test for fexecve at runtime and then decide whether to bind mount or use it based on that.

**Comment by flx42** [comment]
Ok, found the other issue: #2311. But now it's a bigger problem than polluting the rootfs or changing the mtime: it prevents a valid use case.

I'll check if option 1. is fine, if it isn't then we can resurrect the patch I guess.

**Comment by brauner** [comment]
@tych0, @flx42, yes as I said in the prior issue a combination of `fexecve()` and bind-mounting would be preferable. :)

**Comment by flx42** [comment]
@tych0 want to revive your patch then?

**Comment by tych0** [comment]
Yep, I'll try to send it in the next 24 hours.
